daceml/distributed/communication/subarrays.py

```diff
@@ -0,0 +1,355 @@
+import typing
+from typing import Optional, List, Tuple
+
+import sympy as sp
+
+import dace.library
+from dace import SDFG, SDFGState, subsets, symbolic, data
+from dace.transformation import transformation as pm
+from dace.libraries import mpi
+
+from .. import utils as distr_utils
+
+if typing.TYPE_CHECKING:
+    from .node import DistributedMemlet
+
+
+class CommunicationSolverException(Exception):
+    """
+    Exception raised when the communication solver fails to find a solution.
+    """
+    pass
+
+
+MatchedDimensions = List[Tuple[Optional[symbolic.symbol], symbolic.symbol]]
+
+
+def match_subset_axis_to_pgrid(
+        subset: subsets.Range,
+        grid_variables: List[symbolic.symbol]) -> MatchedDimensions:
+    """
+    Matches each axis of the subset to a grid variable if that axis is tiled
+    using that grid variable.
+
+    For example, an axis i*32:(i+1)*32 is matched to the grid variable i with
+    block size 32, and would add (i, 32) to the result.
+
+    If a dimension is not tiled by any variable, the result will contain a None
+    for that dimension.
+
+    :param subset: The subset to match.
+    :param grid_variables: The grid variables to match against.
+    :raises CommunicationSolverException: If matching fails
+    :return: A list of tuples, each containing the grid variable and the
+             matched block size.
+    """
+
+    # avoid import loop
+    from .node import FULLY_REPLICATED_RANK
+
+    blocks = []
+    sizes = subset.size_exact()
+    for i, (start, end, step) in enumerate(subset):
+        if step != 1:
+            raise CommunicationSolverException("Subset step must be 1")
+        bs = sp.Wild("bs", exclude=grid_variables)
+
+        expr = sp.Basic(*map(symbolic.pystr_to_symbolic, (start, end)))
```

Comment:
As far as I understand, it lets me match two expressions at once (start and end). It doesn't seem intelligent about it though; if the first expression matches something that doesn't match the second, it will fail rather than try find another match for the first expression.